cns/client_test.go

```diff
@@ -962,27 +967,57 @@ func TestClient(t *testing.T) {
 		}
 		t.Logf("volume:%q deleted sucessfully", volumeID)
 	}
-	// Test QueryVolumeAsync API
-	queryVolumeAsyncTask, err := cnsClient.QueryVolumeAsync(ctx, queryFilter, querySelection)
-	if err != nil {
-		t.Errorf("Failed to query volumes with QueryVolumeAsync. Error: %+v \n", err)
-		t.Fatal(err)
-	}
-	queryVolumeAsyncTaskInfo, err := GetTaskInfo(ctx, queryVolumeAsyncTask)
-	if err != nil {
-		t.Errorf("Failed to query volumes with QueryVolumeAsync. Error: %+v \n", err)
-		t.Fatal(err)
-	}
-	queryVolumeAsyncTaskResults, err := GetTaskResultArray(ctx, queryVolumeAsyncTaskInfo)
-	if err != nil {
-		t.Errorf("Failed to query volumes with QueryVolumeAsync. Error: %+v \n", err)
-		t.Fatal(err)
+	// Test QueryVolumeAsync API only for vSphere version 7.0.3++
+	if isvSphereVersion70U3orAbove(ctx, c.ServiceContent.About) {
+		queryVolumeAsyncTask, err := cnsClient.QueryVolumeAsync(ctx, queryFilter, querySelection)
+		if err != nil {
+			t.Errorf("Failed to query volumes with QueryVolumeAsync. Error: %+v \n", err)
+			t.Fatal(err)
+		}
+		queryVolumeAsyncTaskInfo, err := GetTaskInfo(ctx, queryVolumeAsyncTask)
+		if err != nil {
+			t.Errorf("Failed to query volumes with QueryVolumeAsync. Error: %+v \n", err)
+			t.Fatal(err)
+		}
+		queryVolumeAsyncTaskResults, err := GetTaskResultArray(ctx, queryVolumeAsyncTaskInfo)
+		if err != nil {
+			t.Errorf("Failed to query volumes with QueryVolumeAsync. Error: %+v \n", err)
+			t.Fatal(err)
+		}
+		for _, queryVolumeAsyncTaskResult := range queryVolumeAsyncTaskResults {
+			queryVolumeAsyncOperationRes := queryVolumeAsyncTaskResult.GetCnsVolumeOperationResult()
+			if queryVolumeAsyncOperationRes.Fault != nil {
+				t.Fatalf("Failed to query volumes with QueryVolumeAsync. fault=%+v", queryVolumeAsyncOperationRes.Fault)
+			}
+			t.Logf("Successfully queried Volume using queryAsync API. queryVolumeAsyncTaskResult: %+v", pretty.Sprint(queryVolumeAsyncTaskResult))
+		}
 	}
-	for _, queryVolumeAsyncTaskResult := range queryVolumeAsyncTaskResults {
-		queryVolumeAsyncOperationRes := queryVolumeAsyncTaskResult.GetCnsVolumeOperationResult()
-		if queryVolumeAsyncOperationRes.Fault != nil {
-			t.Fatalf("Failed to query volumes with QueryVolumeAsync. fault=%+v", queryVolumeAsyncOperationRes.Fault)
+}
+
+// isvSphereVersion70U3orAbove checks if specified version is 7.0 Update 3 or higher
+// The method takes aboutInfo{} as input which contains details about
+// VC version, build number and so on.
+// If the version is 7.0 Update 3 or higher, the method returns true, else returns false
+// along with appropriate errors during failure cases
+func isvSphereVersion70U3orAbove(ctx context.Context, aboutInfo types.AboutInfo) bool {
```

Comment:
I think we ought to move the [govc version helpers](https://github.com/vmware/govmomi/blob//govc/flags/version.go#L51) to another package so you could make use of them here.  There's other use cases for it too.  But this is fine for now.

Reply:
Sounds good. Will follow-up with govc version helpers in another PR